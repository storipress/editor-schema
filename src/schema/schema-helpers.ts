import type { JSONContent } from '@tiptap/core'
import type { Fragment, ParseOptions } from '@tiptap/pm/model'
import { DOMParser, DOMSerializer, Node } from '@tiptap/pm/model'
import { parseToDom, window as win } from '~/utils/environment'
import { DEFAULT_TREE, schema } from './schema'

export const serializer = DOMSerializer.fromSchema(schema)
export const parser = DOMParser.fromSchema(schema)

export function parse(html: string, options?: ParseOptions, domParser: DOMParser = parser): Node {
  const dom = parseToDom(html)
  return domParser.parse(dom, options)
}

export function render(json: Record<string, unknown> = DEFAULT_TREE) {
  const doc = loadDocument(json)
  return renderFragment(doc.content)
}

export interface ArticleMeta {
  clientId: string
  articleId: string
}

export function renderWithMark(node: JSONContent, meta: ArticleMeta) {
  const dom = parseToDom(render(node))
  dom.firstElementChild?.setAttribute('data-sp-article', `${meta.clientId}.${meta.articleId}`)
  dom.lastElementChild?.setAttribute('data-sp-article-end', `${meta.clientId}.${meta.articleId}`)
  return dom.innerHTML
}

export function loadDocument(json: Record<string, unknown> = DEFAULT_TREE) {
  return Node.fromJSON(schema, json)
}

export function renderFragment(fragment: Fragment): string {
  const domFragment = serializer.serializeFragment(fragment, { document: win.document })

  return Array.from(domFragment.childNodes).reduce(
    (result, node) => result + ((node as any).outerHTML || node.nodeValue),
    '',
  )
}
