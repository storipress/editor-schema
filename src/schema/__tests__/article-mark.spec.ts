import { renderWithMark } from '../schema-helpers'

it('can generate html with article mark', () => {
  const html = renderWithMark(
    {
      type: 'doc',
      content: [
        {
          type: 'paragraph',
          content: [
            {
              type: 'text',
              text: 'Line 1',
            },
          ],
        },
        {
          type: 'paragraph',
          content: [
            {
              type: 'text',
              text: 'Line 2',
            },
          ],
        },
      ],
    },
    { articleId: 'article_id', clientId: 'client_id' },
  )

  expect(html).toMatchSnapshot()
})

it('can generate multiple line html with article mark', () => {
  const html = renderWithMark(
    {
      type: 'doc',
      content: [
        {
          type: 'paragraph',
          content: [
            {
              type: 'text',
              text: 'Line 1',
            },
          ],
        },
        {
          type: 'paragraph',
          content: [
            {
              type: 'text',
              text: 'Line 2',
            },
          ],
        },
        {
          type: 'paragraph',
          content: [
            {
              type: 'text',
              text: 'Line 3',
            },
          ],
        },
        {
          type: 'paragraph',
          content: [
            {
              type: 'text',
              text: 'Line 4',
            },
          ],
        },
      ],
    },
    { articleId: 'article_id', clientId: 'client_id' },
  )

  expect(html).toMatchSnapshot()
})
