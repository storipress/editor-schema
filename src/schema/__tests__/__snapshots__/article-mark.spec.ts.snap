// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`can generate html with article mark 1`] = `"<p data-sp-article="client_id.article_id">Line 1</p><p data-sp-article-end="client_id.article_id">Line 2</p>"`;

exports[`can generate multiple line html with article mark 1`] = `"<p data-sp-article="client_id.article_id">Line 1</p><p>Line 2</p><p>Line 3</p><p data-sp-article-end="client_id.article_id">Line 4</p>"`;
